libbeat/processors/actions/rename_fields.go

```diff
@@ -0,0 +1,101 @@
+package actions
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+type renameFields struct {
+	Fields        map[string]string
+	ignoreMissing bool
+	failOnError   bool
+}
+
+func init() {
+	processors.RegisterPlugin("rename_fields",
+		configChecked(newRenameFields,
+			requireFields("fields")))
+}
+
+func newRenameFields(c *common.Config) (processors.Processor, error) {
+	config := struct {
+		Fields        map[string]string `config:"fields"`
+		IgnoreMissing bool              `config:"ignore_missing"`
+		FailOnError   bool              `config:"fail_on_error"`
+	}{}
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the rename_fields configuration: %s", err)
+	}
+
+	f := &renameFields{
+		Fields:        config.Fields,
+		ignoreMissing: config.IgnoreMissing,
+	}
+	return f, nil
+}
+
+func (f *renameFields) Run(event *beat.Event) (*beat.Event, error) {
+
+	var backup common.MapStr
+	// Creates a copy of the event to revert in case of failure
+	if f.failOnError {
+		backup = event.Fields.Clone()
+	}
+
+	for from, to := range f.Fields {
+
+		err := f.renameField(from, to, event.Fields)
+		if err != nil && f.failOnError {
+			logp.Debug("rename", "Revert old event")
+			event.Fields = backup
+			return event, err
+		}
+	}
+
+	return event, nil
+}
+
+func (f *renameFields) renameField(from string, to string, fields common.MapStr) error {
+
+	// Fields cannot be overwritten. Either the target field has to be dropped first or renamed first
+	exists, err := fields.HasKey(to)
+	// TODO: Why is it an error if HasKey is false?
```

Comment:
`HasKey` returns true iff (if and only if) the key exists, so you can ignore the error.

The error is just some extra information to go along with the bool. For example `HasKey` will return an error if the object has a `x.y` and you ask for `x.y.z`, because it expects `y` to be an object, not a scalar.